(RCM-539) Add RC service and client to cluster-agent

ameske opened the pull request:
A work in progress branch for porting the core remote config service into the cluster agent, along with an in-process agent client.

ameske committed:
[RCM] Abstract how RC agent clients retrieve config updates

The "agent client" implementation was hard-coded to only work over a
secure GRPC client provided by a utility package in the datadog core
agent. We have a use case coming up where we want to be able to run both
the core remote config service AND an agent client in the same process
without having to setup and run the secure GRPC server within that
process.

In order to do this, we define an interface for the main config update
request and pull out the GRPC logic into a type that implements this
interface. We provide a convenience constructor to create an RC agent
client backed by the secure GRPC client to prevent the GRPC constructor
from needing to leave the internal package.

ameske committed:
[RCM] Add the core remote config service and a client to the cluster-agent

Adds the core remote config service to the cluster-agent, as well as an
instance of an "agent client" that communicates directly to the core
service instead of over a secure GRPC connection.

This serves as a starting point for any team wanting to build out RC
functionality in the cluster-agent.

ameske commented:
Just for clarification, it's only a WIP because I want to make sure this works for both RC and the team working on the eventual RC integration in the cluster agent before requesting reviews from everybody on the PR. The approach should be as discussed though, so I'm not anticipating any surprises.

ahmed-mez commented on cmd/cluster-agent/app/app.go:
```diff
@@ -194,6 +198,27 @@ func start(cmd *cobra.Command, args []string) error {
+	var rcClient *remote.Client
+	if config.Datadog.GetBool("remote_configuration.enabled") {
+		configService, err = remoteconfig.NewService()
+		if err == nil {
+			if err := configService.Start(context.Background()); err == nil {
+				rcClient, err = remote.NewClient("cluster-agent", configService, version.AgentVersion, []data.Product{data.ProductTesting1}, time.Second*5)
+				if err != nil {
+					log.Errorf("Failed to start local config management service client: %s", err)
+				} else {
+					rcClient.Start()
+				}
+			} else {
+				log.Errorf("Failed to start config management service: %s", err)
+			}
+		} else {
+			log.Errorf("Failed to initialize config management service: %s", err)
+		}
```
Can we [align the happy path to the left](https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88)?
(i.e `if err != nil` instead of if `err == nil`)

ahmed-mez commented on cmd/cluster-agent/app/app.go:
```diff
@@ -194,6 +198,27 @@ func start(cmd *cobra.Command, args []string) error {
+	if config.Datadog.GetBool("remote_configuration.enabled") {
+		configService, err = remoteconfig.NewService()
+		if err == nil {
+			if err := configService.Start(context.Background()); err == nil {
```
```suggestion
			if err := configService.Start(mainCtx); err == nil {
```
I assume `configService.Start` handles a graceful shutdown since it accepts a context?

ahmed-mez commented on cmd/cluster-agent/app/app.go:
```diff
@@ -401,6 +426,8 @@ func start(cmd *cobra.Command, args []string) error {
 		log.Errorf("Error shutdowning metrics server on port %d: %v", metricsPort, err)
 	}
 
+	rcClient.Close()
```
Not sure this is safe if `rcClient` is `nil`

ahmed-mez commented on pkg/config/remote/client.go:
```diff
@@ -52,42 +58,85 @@ type Client struct {
+
+// ClientGetConfigs implements the ConfigUpdater interface for agentGRPCConfigFetcher
+func (g *agentGRPCConfigFetcher) ClientGetConfigs(ctx context.Context, request *pbgo.ClientGetConfigsRequest) (*pbgo.ClientGetConfigsResponse, error) {
+	// This client is running, at the moment, in the trace-agent or the security-agent. The auth token is handled
+	// by the core-agent, running independently. It's not guaranteed it starts before us, or that if it restarts that
+	// the auth token remains the same. Thus we need to do this every request.
```
should update the comment to say it can run in the cluster agent as well?

ameske replied:
I really should drop the reference to the specific agent clients, as this block of code is only relevant for agent clients that communicate with the core service via gRPC (which is not the cluster-agent in this case). I'll do that!

ahmed-mez reviewed (commented):
Thanks @ameske for working on this!

@davidor @Kyle-Verhoog PTAL

ameske committed:
Update cmd/cluster-agent/app/app.go

Co-authored-by: Ahmed Mezghani <38987709+ahmed-mez@users.noreply.github.com>

ameske committed:
Address review

- Refactor error handling on remote config initialization
- Update comment
- Protect against closing a nil rc client

ahmed-mez commented on cmd/cluster-agent/app/app.go:
```diff
@@ -416,3 +433,23 @@ func setupClusterCheck(ctx context.Context) (*clusterchecks.Handler, error) {
+func initializeRemoteConfig(ctx context.Context) (*remote.Client, error) {
+	configService, err := remoteconfig.NewService()
+	if err != nil {
+		return nil, fmt.Errorf("unable to create remote-config service: %v", err)
```
```suggestion
		return nil, fmt.Errorf("unable to create remote-config service: %w", err)
```

ahmed-mez commented on cmd/cluster-agent/app/app.go:
```diff
@@ -416,3 +433,23 @@ func setupClusterCheck(ctx context.Context) (*clusterchecks.Handler, error) {
+	}
+
+	if err := configService.Start(ctx); err != nil {
+		return nil, fmt.Errorf("unable to start remote-config service: %v", err)
```
```suggestion
		return nil, fmt.Errorf("unable to start remote-config service: %w", err)
```

ahmed-mez commented on cmd/cluster-agent/app/app.go:
```diff
@@ -416,3 +433,23 @@ func setupClusterCheck(ctx context.Context) (*clusterchecks.Handler, error) {
+
+	rcClient, err := remote.NewClient("cluster-agent", configService, version.AgentVersion, []data.Product{}, time.Second*5)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create local remote-config client: %v", err)
```
```suggestion
		return nil, fmt.Errorf("unable to create local remote-config client: %w", err)
```

ahmed-mez commented on cmd/cluster-agent/app/app.go:
```diff
@@ -194,6 +198,15 @@ func start(cmd *cobra.Command, args []string) error {
 		log.Debugf("Health check listening on port %d", healthPort)
 	}
 
+	// Initialize remote configuration
+	var rcClient *remote.Client
+	if config.Datadog.GetBool("remote_configuration.enabled") {
+		rcClient, err = initializeRemoteConfig(mainCtx)
+		if err != nil {
+			log.Errorf("Failed to start remote-configuration: %v", err)
+		}
+	}
```
```suggestion
	// Initialize remote configuration
	if config.Datadog.GetBool("remote_configuration.enabled") {
		rcClient, err := initializeRemoteConfig(mainCtx)
		if err != nil {
			log.Errorf("Failed to start remote-configuration: %v", err)
		} else {
		    rcClient.Start()
		    defer rcClient.Close()
		}
	}
```

ahmed-mez commented on cmd/cluster-agent/app/app.go:
```diff
@@ -401,6 +414,10 @@ func start(cmd *cobra.Command, args []string) error {
 		log.Errorf("Error shutdowning metrics server on port %d: %v", metricsPort, err)
 	}
 
+	if rcClient != nil {
+		rcClient.Close()
+	}
+
```
```suggestion
```

ahmed-mez commented on cmd/cluster-agent/app/app.go:
```diff
@@ -416,3 +433,23 @@ func setupClusterCheck(ctx context.Context) (*clusterchecks.Handler, error) {
+		return nil, fmt.Errorf("unable to create local remote-config client: %v", err)
+	}
+
+	rcClient.Start()
```
Nitpick: I'd rather run `rcClient.Start()` in the `start` (caller) function since `initializeRemoteConfig` returns the client anyway, and call `defer rcClient.Close()` right after it.

ameske replied:
That works for me!

The client is primarily returned for product code to register its callbacks on eventually - there aren't any in my PR but your changes that build on this will need the handle to the client to do that. That being said I like making the Start() more deliberate and not hidden in the initialize function.

ahmed-mez reviewed (approved):
A few minor comments left, but LGTM! Thank you Kyle! 🙇

ameske committed:
Update pkg/config/remote/client.go

Co-authored-by: Kyle Verhoog <kyle@verhoog.ca>

ameske committed:
Update cmd/cluster-agent/app/app.go

Co-authored-by: Ahmed Mezghani <38987709+ahmed-mez@users.noreply.github.com>

ameske committed:
Update cmd/cluster-agent/app/app.go

Co-authored-by: Ahmed Mezghani <38987709+ahmed-mez@users.noreply.github.com>

ameske committed:
Update cmd/cluster-agent/app/app.go

Co-authored-by: Ahmed Mezghani <38987709+ahmed-mez@users.noreply.github.com>

ameske committed:
Update cmd/cluster-agent/app/app.go

Co-authored-by: Ahmed Mezghani <38987709+ahmed-mez@users.noreply.github.com>

ameske commented on cmd/cluster-agent/app/app.go:
```diff
@@ -416,3 +433,23 @@ func setupClusterCheck(ctx context.Context) (*clusterchecks.Handler, error) {
+		return nil, fmt.Errorf("unable to start remote-config service: %v", err)
+	}
+
+	rcClient, err := remote.NewClient("cluster-agent", configService, version.AgentVersion, []data.Product{}, time.Second*5)
```
This value is not configurable externally in the core agent at this time. We could use a constant but this is the only place this value is used and I tend to lean on preferring the explicitness in that situation rather than requiring a hop in the file to figure out what the value actually is.

ameske committed:
Update cmd/cluster-agent/app/app.go

Co-authored-by: Ahmed Mezghani <38987709+ahmed-mez@users.noreply.github.com>

ameske committed:
Start rc client in main start func

ogaca-dd reviewed (approved):
Approve for files owned by `agent-shared-components`: cmd/agent/api/grpc.go
